datastore/datastore.go

```diff
@@ -0,0 +1,303 @@
+	ProjectID               string  `json:"project_id"`
+	TokenURI                string  `json:"token_uri"`
+	Type                    string  `json:"type"`
+}
+
+// InventoryData is the data type for inventory data produced by a plugin data
+// source and emitted to the agent's inventory data store
+type InventoryData map[string]interface{}
+
+// MetricData is the data type for events produced by a plugin data source and
+// emitted to the agent's metrics data store
+type MetricData map[string]interface{}
+
+// EventData is the data type for single shot events
+type EventData map[string]interface{}
+
+// NAME - name of plugin
+const NAME string = "datastore"
+
+// ProtocolVersion -
+const ProtocolVersion string = "1"
+
+func Run(log *logrus.Logger, prettyPrint bool, version string) {
+	var data = PluginData{
+		Name:            NAME,
+		ProtocolVersion: ProtocolVersion,
+		PluginVersion:   version,
+		Inventory:       make(map[string]InventoryData),
+		Metrics:         make([]MetricData, 0),
+		Events:          make([]EventData, 0),
+	}
+
+	var keyData KeyData
+
+	keyFile, err := ioutil.ReadFile("/var/secrets/google/key.json")
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = json.Unmarshal(keyFile, &keyData)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "/var/secrets/google/key.json")
+
+	projectID := keyData.ProjectID
+	c := NewClient(projectID)
+
+	//add stackdriver metrics
+	for _, metric := range stackdriverEndpoints {
+		resp, err := getStackdriverResp(c.Sdc, projectID, metric)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		result, err := getStackdriverData(resp)
+		if err != nil {
+			log.Fatal(err)
+		}
+		for _, metricResult := range result {
+			data.Metrics = append(data.Metrics, metricResult)
+		}
+	}
+
+	//add query metrics
+	kinds, err := getDatastoreQueryResult(c.Dsc)
+	if err != nil {
+		log.Fatal(err)
+	}
+	result := getDatastoreData(kinds, projectID)
+
+	for _, metricResult := range result {
+		data.Metrics = append(data.Metrics, metricResult)
+	}
+
+	fatalIfErr(log, helpers.OutputJSON(data, prettyPrint))
+}
+
+func getStackdriverResp(sdl StackDriverImpl, projectID string, metric string) (*monitoring.ListTimeSeriesResponse, error) {
+	startTime := time.Now().UTC().Add(time.Minute * -3)
+	endTime := time.Now().UTC()
+
+	resp, err := sdl.List(projectResource(projectID)).
+		Filter(fmt.Sprintf("metric.type=\"%s\"", metric)).
+		IntervalStartTime(startTime.Format(time.RFC3339)).
+		IntervalEndTime(endTime.Format(time.RFC3339)).
+		Do()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+
+}
+
+func getStackdriverData(resp *monitoring.ListTimeSeriesResponse) ([]map[string]interface{}, error) {
+	var stackdriverMetricBody StackdriverMetric
+	var metricResult []map[string]interface{}
+
+	err := json.Unmarshal(formatResource(resp), &stackdriverMetricBody)
+	if err != nil {
+		return nil, err
+	}
+
+	for id := range stackdriverMetricBody.TimeSeries {
```

Comment:
Rather than getting the iteration as id and calling `stackdriverMetricBody.TimeSeries[id]` a lot below, why not run `for _, item := stackdriverMetricBody.TimeSeries`